sanitycheck: Allow setting of board id

doanac opened the pull request:
The "west flash" commands requires a board-id when more than one
board is attached to the host system. When using a tool like
sanitycheck it will actually hang waiting for stdin to tell it which
board is needed.

This change allows a user to pass -DBOARD_ID=<unique id> to cmake so
that a specific board can be targeted.

Signed-off-by: Andy Doan <andy@foundries.io>

doanac commented:
@mbolivar - FYI. This might be a terrible way to do this, but with this plus using sanitycheck's "--extra-args" flag, I can target a specific board.

nashif commented:
afaik you can invoke west directly in the build directory and specify additional arguments accepted by west. There are way too many options to be used that it does not scale if we translate those to cmake IMO

mbolivar commented:
Maybe we can have sanitycheck use west flash instead of ninja flash, and add a --west-flash--options flag to sanitycheck?

nashif commented:
> Maybe we can have sanitycheck use west flash instead of ninja flash, and add a --west-flash--options flag to sanitycheck?

yeah, I guess this would work

nashif commented:
any updates here?

doanac commented:
Not sure if this is too much of a hack, but here's a way I've found to do this that solves my problem:
~~~
diff --git a/scripts/sanitycheck b/scripts/sanitycheck
index c090ddd896..2981bb1fe7 100755
--- a/scripts/sanitycheck
+++ b/scripts/sanitycheck
@@ -471,12 +471,16 @@ class DeviceHandler(Handler):
     def handle(self):
         out_state = "failed"
 
-        if options.ninja:
-            generator_cmd = "ninja"
+        if options.west_flash_option:
+            command = ["west", "flash", "--skip-rebuild", "-d", self.outdir]
+            command.extend(options.west_flash_option)
         else:
-            generator_cmd = "make"
+            if options.ninja:
+                generator_cmd = "ninja"
+            else:
+                generator_cmd = "make"
 
-        command = [generator_cmd, "-C", self.outdir, "flash"]
+            command = [generator_cmd, "-C", self.outdir, "flash"]
 
         device = options.device_serial
         ser = serial.Serial(
@@ -2681,6 +2685,18 @@ def parse_arguments():
         which will ultimately disable ccache.
         """
     )
+    parser.add_argument(
+        "--west-flash-option", action="append", default=[],
+        help="""Extra options to pass to the "west flash" command when doing
+        running with --device-testing"
+
+        E.g
+         sanitycheck --device-testing --device-serial /dev/ttyACM0 \
+                     --west-flash-option="--board-id=foobar"
+        will translate to
+         west flash -- --board-id=foobar
+        """
+    )
     parser.add_argument("--gcov-tool", default="gcov",
                         help="Path to the gcov tool. Default is gcov in the path.")

~~~

doanac committed:
sanitycheck: Allow custom arguments for west-flash

This allows you to use west to flash the device and pass custom
arguments to the flash command.

Signed-off-by: Andy Doan <andy@foundries.io>

doanac committed:
sanitycheck: Add a "test-only" option

This makes it easy to re-run tests from an existing build directory
w/o trying to rebuild artifacts.

Signed-off-by: Andy Doan <andy@foundries.io>

doanac commented:
I've done a force-push with a different way to handle this.

galak reviewed (dismissed):
I don't know enough about west yet to say, but seems like we're going to probably want more than just --west-flash-option.  I think we should make this `--board-id` at this point and than it can be used for any future things that got plumbed through to west.

mbolivar commented:
> I don't know enough about west yet to say, but seems like we're going to probably want more than just --west-flash-option. I think we should make this `--board-id` at this point and than it can be used for any future things that got plumbed through to west.

I'm not sure I agree with that; --west-flash-option is strictly more general than --board-id. The latter is only supported by some runners. The latter notion of a "board identifier" is, like a flash offset, something that is supported in various ways by different runners through different options whose names come to the present from environment variables which the old shell scripts that the runners were ported from used to respect. Whereas the former stands a decent chance of surviving an attempt to generalize all that.

galak commented:
> I'm not sure I agree with that; --west-flash-option is strictly more general than --board-id. The latter is only supported by some runners. The latter notion of a "board identifier" is, like a flash offset, something that is supported in various ways by different runners through different options whose names come to the present from environment variables which the old shell scripts that the runners were ported from used to respect. Whereas the former stands a decent chance of surviving an attempt to generalize all that.

Fair point, so should the option just be more generic as `--west-options` so it can apply to cases beyond just flash?

doanac commented:
> so should the option just be more generic as `--west-options` so it can apply to cases beyond just flash?

Marti could better answer, but I don't think there's a good intersection of common options that would be useful. ie - the common options like "builddir" will already be known to sanitycheck. And things like "board-id" are only applicable to certain west sub-commands eg "flash".

This isn't the right time, but I think long term Zephyr ought to consider splitting up sanitycheck into 2 tools. One for building and one for test execution. That might simplify conversations like this.

mbolivar commented:
> > so should the option just be more generic as `--west-options` so it can apply to cases beyond just flash?
> 
> Marti could better answer, but I don't think there's a good intersection of common options that would be useful. ie - the common options like "builddir" will already be known to sanitycheck. And things like "board-id" are only applicable to certain west sub-commands eg "flash".

Right.

The only common west options are things like `--verbose` , `--zephyr-base`, etc. Those have to come before the command name in the west argument list (i.e. it's `west [common-opts] <command-name> [command-opts]`), so adding `--west-options` wouldn't preclude the need for `--west-flash-options`.

The flash options are what we care about for device testing in sanitycheck. Adding `--west-flash-options` is admittedly yet another "organic" evolution to sanitycheck's device testing feature, which itself is an organic outgrowth from past features. At some point we need to discuss rationalizing all of this after gathering use cases for sanitycheck now that we've been using it for a few years, but I don't know if that time is now.

> 
> This isn't the right time, but I think long term Zephyr ought to consider splitting up sanitycheck into 2 tools. One for building and one for test execution. That might simplify conversations like this.

This I think is a good line item to add to the use case list for 'what would sanitycheck 2.0 really look like' as described above.

doanac commented:
You guys think this is merge-worthy? We've been using these patches for a few weeks now to do bare-metal testing of zephyr's master.

SebastianBoe commented on scripts/sanitycheck:
```diff
@@ -2883,6 +2887,18 @@ def parse_arguments():
+                     --west-flash-option="--board-id=foobar"
+        will translate to
+         west flash -- --board-id=foobar
+        """
```
Shouldn't there be two options?

One option for using west, and another for appending options to it.

What if you want to use west, but don't want any additional options?

doanac replied:
@SebastianBoe That seems reasonable. 
@nashif  @mbolivar  - does that make sense to you guys?

mbolivar replied:
I guess it makes sense? I'm not really sure though; `ninja flash` is exactly "west without additional options", by design...

mbolivar replied:
clarification: of you want to do it that way, no objections from me, but it seems a bit strange for the above reason.

marc-hb replied:
```python
   "--west-flash [ west_flash_options ]"
help="uses west instead of ninja or make to flash"
```

doanac replied:
I'm not sure how to interpret @marc-hb's comment. Are you saying you want me to rename the existing flag from --west-flash-option -> --west-flash. Remove the old option in favor of the method you are showing here?(which isn't exactly python argparse friendly). or make "--west-flash" a "store_true" option that then explicitly uses west-flash and any options we may have added via --west-flash-option ?

marc-hb replied:
I'm suggesting: `"--west-flash", nargs=?, default=None, const=[]`

(`default` can probably be omitted)

marc-hb commented on scripts/sanitycheck:
```diff
@@ -2792,9 +2792,14 @@ def parse_arguments():
         help="Save list of tests to be run to file.")
 
-    parser.add_argument(
+    group = parser.add_mutually_exclusive_group()
```
```python
test_xor_build = parser.add_mutually_exclusive_group()
```
... or any more specific name than "group" so add_mutually_exclusive_group() can be used again elsewhere if needed.

marc-hb commented on scripts/sanitycheck:
```diff
@@ -644,12 +644,16 @@ class DeviceHandler(Handler):
+                generator_cmd = "make"
 
-        command = [generator_cmd, "-C", self.outdir, "flash"]
+            command = [generator_cmd, "-C", self.outdir, "flash"]
```
`logging.debug(command)` or equivalent

marc-hb commented:
I asked @mbolivar a (naive) question about west and sanitycheck. With hindsight my question was not far from: "Is there some kind of 'west check'" subcommand to invoke sanitycheck in fun ways?" I wondered about that because I read west is the one front end to rule them all, even the least related things.

So @mbolivar naturally thought of and pointed me at this counter-example where - the other way round - sanitycheck invokes west...

>  Adding --west-flash-options is admittedly yet another "organic" evolution to sanitycheck's device testing feature, which itself is an organic outgrowth from past features. At some point we need to discuss rationalizing all of this after gathering use cases for sanitycheck

I think good old boxes and arrows on a whiteboard showing the graph and "vision" of which component is ideally supposed to invoke what in the long term would be a good start. Hopefully a graph without any circle ;-)

My 2 cents and sorry if that diagram has been made already, if yes please share.

doanac commented:
rebased on current master and addressed 2 of the 3 review comments. (logging.debug and using a better variable name for the mutually exclusive argument group)

doanac commented:
okay - latest git-push -f should have things in shape. I've tested the 2 commits in this PR here:

 https://ci.foundries.io/projects/zephyr/builds/5286/sanity-test-nrf52/artifacts/console.log

paulgavrikov commented:
Are there any plans to merge this into the master?

galak commented:
> Are there any plans to merge this into the master?

Needs review / approval from @SebastianBoe

galak commented:
@paulgavrikov not sure if you'd have any interest in looking to add board id support to west?
